lib/ecto/adapters/mysql/connection.ex

```diff
@@ -0,0 +1,502 @@
+      "false"
+    end
+
+    defp expr({:in, _, [left, right]}, sources) when is_list(right) do
+      args = Enum.map_join right, ",", &expr(&1, sources)
+      expr(left, sources) <> " IN (" <> args <> ")"
+    end
+
+    defp expr({:in, _, [left, {:^, _, [ix, length]}]}, sources) do
+      args = Enum.map_join ix+1..ix+length, ",", &"$#{&1}"
+      expr(left, sources) <> " IN (" <> args <> ")"
+    end
+
+    defp expr({:is_nil, _, [arg]}, sources) do
+      "#{expr(arg, sources)} IS NULL"
+    end
+
+    defp expr({:not, _, [expr]}, sources) do
+      "NOT (" <> expr(expr, sources) <> ")"
+    end
+
+    defp expr({:fragment, _, parts}, sources) do
+      Enum.map_join(parts, "", fn
+        part when is_binary(part) -> part
+        expr -> expr(expr, sources)
+      end)
+    end
+
+    defp expr({fun, _, args}, sources) when is_atom(fun) and is_list(args) do
+      case handle_call(fun, length(args)) do
+        {:binary_op, op} ->
+          [left, right] = args
+          op_to_binary(left, sources) <>
+          " #{op} "
+          <> op_to_binary(right, sources)
+
+        {:fun, fun} ->
+          "#{fun}(" <> Enum.map_join(args, ", ", &expr(&1, sources)) <> ")"
+      end
+    end
+
+    defp expr(list, sources) when is_list(list) do
+      "ARRAY[" <> Enum.map_join(list, ",", &expr(&1, sources)) <> "]"
+    end
+
+    defp expr(%Ecto.Query.Tagged{value: other, type: type}, sources) do
+      expr(other, sources) <> "::" <> ecto_to_db(type)
+    end
+
+    defp expr(nil, _sources),   do: "NULL"
+    defp expr(true, _sources),  do: "TRUE"
+    defp expr(false, _sources), do: "FALSE"
+
+    defp expr(literal, _sources) when is_binary(literal) do
+      "'#{escape_string(literal)}'"
+    end
+
+    defp expr(literal, _sources) when is_integer(literal) do
+      String.Chars.Integer.to_string(literal)
+    end
+
+    defp expr(literal, _sources) when is_float(literal) do
+      String.Chars.Float.to_string(literal) <> "::float"
+    end
+
+    defp op_to_binary({op, _, [_, _]} = expr, sources) when op in @binary_ops do
+      "(" <> expr(expr, sources) <> ")"
+    end
+
+    defp op_to_binary(expr, sources) do
+      expr(expr, sources)
+    end
+
+    defp create_names(query) do
+      sources = query.sources |> Tuple.to_list
+      Enum.reduce(sources, [], fn {table, model}, names ->
+        name = unique_name(names, String.first(table), 0)
+        [{table, name, model}|names]
+      end) |> Enum.reverse |> List.to_tuple
+    end
+
+    # Brute force find unique name
+    defp unique_name(names, name, counter) do
+      counted_name = name <> Integer.to_string(counter)
+      if Enum.any?(names, fn {_, n, _} -> n == counted_name end) do
+        unique_name(names, name, counter + 1)
+      else
+        counted_name
+      end
+    end
+
+    ## DDL
+
+    alias Ecto.Migration.Table
+    alias Ecto.Migration.Index
+    alias Ecto.Migration.Reference
+    def ddl_exists(%Table{name: name}) do
```

Comment:
Btw, the reason you didn't add the ddl exists for indexes is because we don't have unit tests for it... because the unit test would be to just copy the SQL posted here... which doesn't make sense. So don't worry... you would have caught it in the integration test anyway. :)